rbx_types/src/terrain.rs

```diff
@@ -0,0 +1,548 @@
+        let water_occupancy: u8 = (self.water_occupancy * 255.0) as u8;
+        (solid_occupancy, water_occupancy)
+    }
+
+    fn encode_run_length(&self, count: u16) -> Vec<u8> {
+        assert!(
+            (1..=256).contains(&count),
+            "Invalid voxel RLE count: {}",
+            count
+        );
+
+        let (solid_occupancy, water_occupancy) = self.get_encode_data();
+        let mut flag = self.material as u8;
+        let mut to_write: Vec<u8> = vec![];
+
+        if solid_occupancy != 0xFF && solid_occupancy != 0x00 {
+            // Should we store the solid occupancy value?
+            flag |= 0b01000000;
+            to_write.push(solid_occupancy);
+        }
+        if count > 1 || water_occupancy != 0 {
+            // Should we store the count (amount of voxels this run length) value?
+            flag |= 0b10000000;
+            if water_occupancy == 0 {
+                to_write.push((count - 1) as u8);
+            } else {
+                to_write.push(0);
+                to_write.push(water_occupancy);
+            }
+        }
+        to_write.insert(0, flag);
+
+        if water_occupancy != 0x00 && count > 1 {
+            /* Shorelines uses a new water occupancy value in the voxel data. Because of this,
+            Roblox uses a hack to avoid having to reduce their 6 bits of material ID freedom
+            by writing voxels with a count bit set to 1 and no count. This means we have to write
+            all voxels in the run length manually. */
+            return to_write
+                .iter()
+                .cycle()
+                .copied()
+                .take(to_write.len() * count as usize)
+                .collect();
+        }
+        to_write
+    }
+
+    /// Sets occupancy data for a `Voxel`. Water occupancy is from the
+    /// new Shorelines feature. Occupancy values are between `0.0` and `1.0`,
+    /// as a percentage of the voxel.
+    pub fn set_occupancy(&mut self, solid_occupancy: f32, water_occupancy: f32) {
+        let solid_occupancy = solid_occupancy.clamp(0.0, 1.0);
+        let water_occupancy = water_occupancy.clamp(0.0, 1.0);
+
+        // If we have nothing in there, it should just be air.
+        if solid_occupancy == 0.0 && water_occupancy == 0.0 {
+            self.solid_occupancy = 1.0;
+            self.water_occupancy = 0.0;
+            self.material = Material::Air;
+            return;
+        }
+
+        // We should encode water as a normal, non-shorelines voxel if there's no solids.
+        if (solid_occupancy == 0.0 || self.material == Material::Air) && water_occupancy > 0.0 {
+            self.solid_occupancy = water_occupancy;
+            self.water_occupancy = 0.0;
+            self.material = Material::Water;
+            return;
+        }
+
+        self.solid_occupancy = solid_occupancy;
+
+        // Full with a solid (non-air) material? We can't have any water.
+        if solid_occupancy == 1.0 && water_occupancy > 0.0 {
```

Comment:
```suggestion
        if solid_occupancy == 1.0 {
```
I don't think we actually need to check the water occupancy here.